express_relay/sdk/python/express_relay_utils/express_relay_client.py

```diff
@@ -0,0 +1,404 @@
+        params = {
+            "chain_ids": chain_ids,
+        }
+        await self.send_ws_message("subscribe", params)
+
+    async def unsubscribe_chains(self, chain_ids: list[str]):
+        """
+        Unsubscribes websocket from a list of chain IDs for new opportunities.
+
+        Args:
+            chain_ids: A list of chain IDs to unsubscribe from.
+        """
+        params = {
+            "chain_ids": chain_ids,
+        }
+        await self.send_ws_message("unsubscribe", params)
+
+    async def submit_bid(
+        self, bid: Bid, subscribe_to_updates: bool = True, **kwargs
+    ) -> UUID:
+        """
+        Submits a bid to the auction server.
+
+        Args:
+            bid: An object representing the bid to submit.
+            subscribe_to_updates: A boolean indicating whether to subscribe to the bid status updates.
+            kwargs: Keyword arguments to pass to the HTTP post request.
+        Returns:
+            The ID of the submitted bid.
+        """
+        bid_dict = bid.model_dump()
+        if subscribe_to_updates:
+            result = await self.send_ws_message("post_bid", bid_dict)
+            bid_id = UUID(result.get("id"))
+        else:
+            async with httpx.AsyncClient() as client:
+                resp = await client.post(
+                    urllib.parse.urlparse(self.server_url)
+                    ._replace(path="/v1/bids")
+                    .geturl(),
+                    json=bid_dict,
+                    **kwargs,
+                )
+
+            resp.raise_for_status()
+            bid_id = UUID(resp.json().get("id"))
+
+        return bid_id
+
+    async def submit_opportunity_bid(
+        self,
+        opportunity_bid: OpportunityBid,
+        subscribe_to_updates: bool = True,
+        **kwargs,
+    ) -> UUID:
+        """
+        Submits a bid on an opportunity to the server via websocket.
+
+        Args:
+            opportunity_bid: An object representing the bid to submit on an opportunity.
+            subscribe_to_updates: A boolean indicating whether to subscribe to the bid status updates.
+            kwargs: Keyword arguments to pass to the HTTP post request.
+        Returns:
+            The ID of the submitted bid.
+        """
+        opportunity_bid_dict = opportunity_bid.model_dump()
+        if subscribe_to_updates:
+            result = await self.send_ws_message(
+                "post_opportunity_bid",
+                {
+                    "opportunity_id": opportunity_bid.opportunity_id,
+                    "amount": opportunity_bid.amount,
+                    "executor": opportunity_bid.executor,
+                    "permission_key": opportunity_bid.permission_key,
+                    "signature": opportunity_bid.signature,
+                    "valid_until": opportunity_bid.valid_until,
+                },
+            )
+            bid_id = UUID(result.get("id"))
+        else:
+            async with httpx.AsyncClient() as client:
```

Comment:
I searched a bit and defining a new client per method call doesn't seem like a good idea:
https://www.python-httpx.org/advanced/clients/
I suggest creating a single client in the constructor and use it here. Also we can accept the httpx client configuration in the constructor and set it once and for all the methods.

Reply:
This is because the link does not assume we are doing this in multiple functions. Each time the user call this functions, a new client is instantiated which is suboptimal:
https://www.python-httpx.org/async/#opening-and-closing-clients